Add support for pluzz.francetv.fr

melmorabity opened the pull request:
This plugin adds support for the French public TV video platform Pluzz. It supports both VOD and live streams. Some streams may be geo-locked to France. The platform currently provides HDS, HLS and DASH streams.

Some URLs for testing:
* live
  * http://pluzz.francetv.fr/france2 (may be geo-locked at some hours)
  * http://pluzz.francetv.fr/franceinfo (not geo-locked)
* VOD
  * http://pluzz.francetv.fr/videos/jt_1920.html (not geo-locked)
  * http://pluzz.francetv.fr/videos/sherlock.html (geo-locked)
  * any video on http://pluzz.francetv.fr/depuis-l-etranger/france3 (not geo-locked)

I have anyway some issues when selecting HDS streams with this plugin: I currently can't tell if it's related to my plugin implementation or to streamlink:
* No sound when trying to play HDS version with this URL (HLS is OK):
```
$ streamlink -Q http://pluzz.francetv.fr/franceinfo 576p_hds -O | ffprobe pipe:0
ffprobe version 3.1.6 Copyright (c) 2007-2016 the FFmpeg developers
  built with gcc 6.2.1 (GCC) 20160916 (Red Hat 6.2.1-2)
  configuration: --prefix=/usr --bindir=/usr/bin --datadir=/usr/share/ffmpeg --incdir=/usr/include/ffmpeg --libdir=/usr/lib64 --mandir=/usr/share/man --arch=x86_64 --optflags='-O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -m64 -mtune=generic' --extra-ldflags='-Wl,-z,relro -specs=/usr/lib/rpm/redhat/redhat-hardened-ld' --enable-bzlib --disable-crystalhd --enable-fontconfig --enable-frei0r --enable-gcrypt --enable-gnutls --enable-ladspa --enable-libass --enable-libbluray --enable-libcdio --enable-indev=jack --enable-libfreetype --enable-libfribidi --enable-libgsm --enable-libmp3lame --enable-nvenc --extra-cflags=-I/usr/include/nvenc --enable-openal --enable-opencl --enable-opengl --enable-libopenjpeg --enable-libopus --enable-libpulse --enable-libschroedinger --enable-libsoxr --enable-libspeex --enable-libtheora --enable-libvorbis --enable-libv4l2 --enable-libvpx --enable-libx264 --enable-libx265 --enable-libxvid --enable-x11grab --enable-avfilter --enable-avresample --enable-postproc --enable-pthreads --disable-static --enable-shared --enable-gpl --disable-debug --disable-stripping --shlibdir=/usr/lib64 --enable-libmfx --enable-runtime-cpudetect
  libavutil      55. 28.100 / 55. 28.100
  libavcodec     57. 48.101 / 57. 48.101
  libavformat    57. 41.100 / 57. 41.100
  libavdevice    57.  0.101 / 57.  0.101
  libavfilter     6. 47.100 /  6. 47.100
  libavresample   3.  0.  0 /  3.  0.  0
  libswscale      4.  1.100 /  4.  1.100
  libswresample   2.  1.100 /  2.  1.100
  libpostproc    54.  0.100 / 54.  0.100
[flv @ 0x55835cd85780] video stream discovered after head already parsed
[h264 @ 0x55835cd86ac0] Increasing reorder buffer to 1
Input #0, flv, from 'pipe:0':
  Duration: N/A, start: 0.040000, bitrate: N/A
    Stream #0:0: Video: h264 (Main), yuv420p(tv, bt470bg), 1024x576 [SAR 1:1 DAR 16:9], 25 fps, 25 tbr, 1k tbn, 50 tbc
```
* Download issues with some VOD streams (HLS is OK; some others streams like http://pluzz.francetv.fr/videos/jt_1920.html are OK in HDS):
```
$ streamlink http://pluzz.francetv.fr/videos/sherlock.html 1466k
[cli][info] Found matching plugin pluzz for URL http://pluzz.francetv.fr/videos/sherlock.html
[cli][info] Available streams: 396p_alt, 144p_alt, 288p_alt, 180p_alt, 180k (worst), 303k, 144p, 180p, 575k, 288p, 831k, 396p, 1466k, 576p (best)
[cli][info] Opening stream: 1466k (hds)
[stream.hds][error] Failed to open fragment 1-1: Unable to open URL: http://replayftv-vh.akamaihd.net/z/streaming-adaptatif_france-dom-tom/2016/S52/J3/150839881-5863c043b588e-,standard1,standard2,standard3,standard4,standard5,.mp4.csmil/4_8d03811cc2126a84_Seg1-Frag1 (403 Client Error: Forbidden for url: http://replayftv-vh.akamaihd.net/z/streaming-adaptatif_france-dom-tom/2016/S52/J3/150839881-5863c043b588e-,standard1,standard2,standard3,standard4,standard5,.mp4.csmil/4_8d03811cc2126a84_Seg1-Frag1?hdntl=exp%3D1483408945~acl%3D%2Fz%2Fstreaming-adaptatif_france-dom-tom%2F2016%2FS52%2FJ3%2F150839881-5863c043b588e-%2Cstandard1%2Cstandard2%2Cstandard3%2Cstandard4%2Cstandard5%2C.mp4.csmil%2A~data%3Dhdntl~hmac%3D5bfd9ee870e723a0e91271f5c1c4cbefe665545c04e9924b9c482ed4a4050014&hdcore=3.1.0&pvtoken=st%3D0~exp%3D9999999999~acl%3D%2A~data%3DZXhwPTE0ODM0MDg5NDV%2BYWNsPSUyZip%2BZGF0YT1wdmMsc35obWFjPWM2NWUzM2U1YzYwNzFkZDRlNTcwMWE2MTM2Y2E0N2RmM2U3YzhjYjg4N2UwMTdmZTNmMjNkODZhYzRiZmNjNDk%3D%21VMDyitlpwrkBYLvGBUZHtLl7f%2FlHc1eEiIMQg0EOxC8%3D~hmac%3Dd512256c87d3310cc74bdaec9c6a3014a1e17a40c83ff7fa15b4ef836116a033)
[stream.hds][error] Failed to open fragment 1-1: Unable to open URL: http://replayftv-vh.akamaihd.net/z/streaming-adaptatif_france-dom-tom/2016/S52/J3/150839881-5863c043b588e-,standard1,standard2,standard3,standard4,standard5,.mp4.csmil/4_8d03811cc2126a84_Seg1-Frag1 (403 Client Error: Forbidden for url: http://replayftv-vh.akamaihd.net/z/streaming-adaptatif_france-dom-tom/2016/S52/J3/150839881-5863c043b588e-,standard1,standard2,standard3,standard4,standard5,.mp4.csmil/4_8d03811cc2126a84_Seg1-Frag1?hdntl=exp%3D1483408945~acl%3D%2Fz%2Fstreaming-adaptatif_france-dom-tom%2F2016%2FS52%2FJ3%2F150839881-5863c043b588e-%2Cstandard1%2Cstandard2%2Cstandard3%2Cstandard4%2Cstandard5%2C.mp4.csmil%2A~data%3Dhdntl~hmac%3D5bfd9ee870e723a0e91271f5c1c4cbefe665545c04e9924b9c482ed4a4050014&hdcore=3.1.0&pvtoken=st%3D0~exp%3D9999999999~acl%3D%2A~data%3DZXhwPTE0ODM0MDg5NDV%2BYWNsPSUyZip%2BZGF0YT1wdmMsc35obWFjPWM2NWUzM2U1YzYwNzFkZDRlNTcwMWE2MTM2Y2E0N2RmM2U3YzhjYjg4N2UwMTdmZTNmMjNkODZhYzRiZmNjNDk%3D%21VMDyitlpwrkBYLvGBUZHtLl7f%2FlHc1eEiIMQg0EOxC8%3D~hmac%3Dd512256c87d3310cc74bdaec9c6a3014a1e17a40c83ff7fa15b4ef836116a033)
[stream.hds][error] Failed to open fragment 1-1: Unable to open URL: http://replayftv-vh.akamaihd.net/z/streaming-adaptatif_france-dom-tom/2016/S52/J3/150839881-5863c043b588e-,standard1,standard2,standard3,standard4,standard5,.mp4.csmil/4_8d03811cc2126a84_Seg1-Frag1 (403 Client Error: Forbidden for url: http://replayftv-vh.akamaihd.net/z/streaming-adaptatif_france-dom-tom/2016/S52/J3/150839881-5863c043b588e-,standard1,standard2,standard3,standard4,standard5,.mp4.csmil/4_8d03811cc2126a84_Seg1-Frag1?hdntl=exp%3D1483408945~acl%3D%2Fz%2Fstreaming-adaptatif_france-dom-tom%2F2016%2FS52%2FJ3%2F150839881-5863c043b588e-%2Cstandard1%2Cstandard2%2Cstandard3%2Cstandard4%2Cstandard5%2C.mp4.csmil%2A~data%3Dhdntl~hmac%3D5bfd9ee870e723a0e91271f5c1c4cbefe665545c04e9924b9c482ed4a4050014&hdcore=3.1.0&pvtoken=st%3D0~exp%3D9999999999~acl%3D%2A~data%3DZXhwPTE0ODM0MDg5NDV%2BYWNsPSUyZip%2BZGF0YT1wdmMsc35obWFjPWM2NWUzM2U1YzYwNzFkZDRlNTcwMWE2MTM2Y2E0N2RmM2U3YzhjYjg4N2UwMTdmZTNmMjNkODZhYzRiZmNjNDk%3D%21VMDyitlpwrkBYLvGBUZHtLl7f%2FlHc1eEiIMQg0EOxC8%3D~hmac%3Dd512256c87d3310cc74bdaec9c6a3014a1e17a40c83ff7fa15b4ef836116a033)
[stream.hds][error] Failed to open fragment 1-2: Unable to open URL: http://replayftv-vh.akamaihd.net/z/streaming-adaptatif_france-dom-tom/2016/S52/J3/150839881-5863c043b588e-,standard1,standard2,standard3,standard4,standard5,.mp4.csmil/4_8d03811cc2126a84_Seg1-Frag2 (403 Client Error: Forbidden for url: http://replayftv-vh.akamaihd.net/z/streaming-adaptatif_france-dom-tom/2016/S52/J3/150839881-5863c043b588e-,standard1,standard2,standard3,standard4,standard5,.mp4.csmil/4_8d03811cc2126a84_Seg1-Frag2?hdntl=exp%3D1483408945~acl%3D%2Fz%2Fstreaming-adaptatif_france-dom-tom%2F2016%2FS52%2FJ3%2F150839881-5863c043b588e-%2Cstandard1%2Cstandard2%2Cstandard3%2Cstandard4%2Cstandard5%2C.mp4.csmil%2A~data%3Dhdntl~hmac%3D5bfd9ee870e723a0e91271f5c1c4cbefe665545c04e9924b9c482ed4a4050014&hdcore=3.1.0&pvtoken=st%3D0~exp%3D9999999999~acl%3D%2A~data%3DZXhwPTE0ODM0MDg5NDV%2BYWNsPSUyZip%2BZGF0YT1wdmMsc35obWFjPWM2NWUzM2U1YzYwNzFkZDRlNTcwMWE2MTM2Y2E0N2RmM2U3YzhjYjg4N2UwMTdmZTNmMjNkODZhYzRiZmNjNDk%3D%21VMDyitlpwrkBYLvGBUZHtLl7f%2FlHc1eEiIMQg0EOxC8%3D~hmac%3Dd512256c87d3310cc74bdaec9c6a3014a1e17a40c83ff7fa15b4ef836116a033)
...
```

beardypig commented on src/streamlink/plugins/pluzz.py:
```diff
@@ -0,0 +1,122 @@
+            if '.f4m' in video_url and swf_url is not None:
+                res = http.get(self.HDS_TOKEN_URL.format(video_url))
+                video_url = res.text
+                for stream in HDSStream.parse_manifest(self.session, video_url, pvswf=swf_url).items():
```
Have you tried using `AkamaiHDStream` for troublesome streams? They might have extra auth/verification.

beardypig commented:
Looks really good, mega +1 for the tests. I haven't been able to test it properly, but the code looks solid. Thanks :-)

melmorabity commented:
I've just added a fix, for Python 3, when checking video availability.

About the download issues with some VOD streams in HLS, it looks like such streams are not supported by the algorithm implemented in HDSStream.
For example, with the following data (thanks Wireshark):
* pv-2.0 token, from manifest: `ZXhwPTE0ODM1MzE0NTZ+YWNsPSUyZip+ZGF0YT1wdmMsc35obWFjPWQ5YTZiNzUxNGRkMWYyNmFkYTRhYmExZDE4ODJhMmE0Yjg0NzIxZjgxMWJmMzFkYjdmM2Y2NzkxMzk0ODBkZGU=;hdntl=exp=1483531456~acl=%2fz%2fstreaming-adaptatif_france-dom-tom%2f2017%2fS01%2fJ1%2f151177312-586b3e5125c81-,standard1,standard2,standard3,standard4,standard5,qaa,.mp4.csmil*~data=hdntl~hmac=e3c4600d59a16fee05aedd44cfa9ec45515f255e42288c7c9850827add7e91cd`
* pvswf: `http://staticftv-a.akamaihd.net/player/bower_components/player_flash/dist/FranceTVNVPVFlashPlayer.akamai-05bcfb8a5a233afa06de66e8c03d6642.swf`

The pvtoken argument generated by HLSStream is:
```
st=0~exp=9999999999~acl=*~data=ZXhwPTE0ODM1MzE0NTZ+YWNsPSUyZip+ZGF0YT1wdmMsc35obWFjPWQ5YTZiNzUxNGRkMWYyNmFkYTRhYmExZDE4ODJhMmE0Yjg0NzIxZjgxMWJmMzFkYjdmM2Y2NzkxMzk0ODBkZGU=!VMDyitlpwrkBYLvGBUZHtLl7f/lHc1eEiIMQg0EOxC8=~hmac=831ac38ffb76356aca35c256bdb7537a4616e3e1b34ddb07990dff39df2866ec
```
whereas the one I actually got in my browser is:
```
exp=9999999999~acl=%2f%2a~data=ZXhwPTE0ODM1MzE0NTZ+YWNsPSUyZip+ZGF0YT1wdmMsc35obWFjPWQ5YTZiNzUxNGRkMWYyNmFkYTRhYmExZDE4ODJhMmE0Yjg0NzIxZjgxMWJmMzFkYjdmM2Y2NzkxMzk0ODBkZGU=!VMDyitlpwrkBYLvGBUZHtLl7f/lHc1eEiIMQg0EOxC8=~hmac=6D58DAE5B55906D47D0A40301EA40E9E21BFC7F9503479094AAC702C113FD57B
```
Notice that the swf hash is the same in both outputs.
The hdcore version for pluzz.francetv.fr is 3.10.0, by the way.
If the pv-2.0 doesn't contain data (i.e. only contains the hdntl parameter), it works. As a result I've modified the plugin to workaround such cases (ignore streams with data in their manifest pv-2.0).

beardypig commented:
Looks good to merge to me.

beardypig commented:
Thanks @melmorabity :)
